src/jedi_nft.cairo

```diff
@@ -0,0 +1,459 @@
+    }
+
+
+    #[external(v0)]
+    impl JediNFTImpl of super::IJediNFT<ContractState> {
+        fn tokenURI(self: @ContractState, token_id: u256) -> Span<felt252> {
+            let base_uri = self._uri.read();
+            let new_base_uri: Array<felt252> = base_uri.snapshot.clone();
+            let mut tmp = InternalImpl::append_number_ascii(new_base_uri, token_id);
+            tmp.append('.json');
+            return tmp.span();
+        }
+
+        fn contractURI(self: @ContractState) -> Span<felt252> {
+            return self._contract_uri.read();
+        }
+
+        fn is_completed(self: @ContractState, task_id: u128, address: ContractAddress) -> bool {
+            return self._completed_tasks.read((task_id, address));
+        }
+
+        fn set_merkle_root(ref self: ContractState, task_id: u128, merkle_root: felt252) {
+            self._only_owner();
+            self._merkle_roots.write(task_id, merkle_root);
+        }
+
+        fn get_merkle_root(self: @ContractState, task_id: u128) -> felt252 {
+            return self._merkle_roots.read(task_id);
+        }
+
+        fn mint_whitelist(
+            ref self: ContractState,
+            token_id: u128,
+            proof: Array<felt252>,
+            token_metadata: TokenMetadata
+        ) {
+            let caller = get_caller_address();
+            let merkle_root = self._merkle_roots.read(token_metadata.task_id);
+            assert(merkle_root != 0, 'merkle root not set');
+            let mut merkle_tree: MerkleTree<Hasher> = MerkleTreeTrait::new();
+            let mut leaf: felt252 = merkle_tree.hasher.hash(caller.into(), token_id.into());
+            leaf = merkle_tree.hasher.hash(leaf, token_metadata.task_id.into());
+            leaf = merkle_tree.hasher.hash(leaf, token_metadata.name.into());
+            leaf = merkle_tree.hasher.hash(leaf, token_metadata.rank.into());
+            leaf = merkle_tree.hasher.hash(leaf, token_metadata.score.into());
+            leaf = merkle_tree.hasher.hash(leaf, token_metadata.level.into());
+            leaf = merkle_tree.hasher.hash(leaf, token_metadata.total_eligible_users.into());
+
+            let result = merkle_tree.verify(merkle_root, leaf, proof.span());
+            assert(result == true, 'verify failed');
+
+            self._token_metadata.write(token_id, token_metadata);
+            let is_minted = self._completed_tasks.read((token_metadata.task_id, caller));
+            assert(!is_minted, 'ALREADY_MINTED');
+            self._completed_tasks.write((token_metadata.task_id, caller), true);
+            let mut erc721_self = ERC721::unsafe_new_contract_state();
+            erc721_self._mint(to: caller, token_id: token_id.into());
+        }
+
+        fn get_token_metadata(self: @ContractState, token_id: u128) -> TokenMetadata {
+            return self._token_metadata.read(token_id);
+        }
+
+        fn set_mint_sig_pub_key(ref self: ContractState, mint_sig_public_key: felt252) {
```

Comment:
have you consider adding event for `_mint_sig_public_key` and `_merkle_roots` write?